crates/fuzzing/wasm-spec-interpreter/build.rs

```diff
@@ -0,0 +1,95 @@
+/// Build the OCaml code and statically link it into the Rust library; see the [ocaml-interop
+/// example](https://github.com/tezedge/ocaml-interop/blob/master/testing/rust-caller/build.rs) for
+/// more details. After playing with this a bit, I discovered that the best approach to avoid
+/// missing symbols was to imitate `dune`: I observed `rm -rf _build && dune build
+/// ./ocaml/interpret.exe.o --display=verbose` and used that as a pattern, now encoded in
+/// `ocaml/Makefile` for easier debugging.
+use std::{env, path::PathBuf, process::Command};
+
+const LIB_NAME: &'static str = "interpret";
+const OCAML_DIR: &'static str = "ocaml";
+const HAS_LIB_FEATURE: &'static str = "has-libinterpret";
+
+fn main() {
+    let out_dir = &env::var("OUT_DIR").unwrap();
+
+    // Re-run if changed.
+    println!("cargo:rerun-if-changed={}/{}.ml", OCAML_DIR, LIB_NAME);
+    println!("cargo:rerun-if-changed={}/Makefile", OCAML_DIR);
+
+    if let Some(other_dir) = env::var_os("FFI_LIB_DIR") {
+        // Link with the given library. (TODO might need to copy these in to
+        // OUT_DIR)
+        println!("cargo:rustc-link-search={}", other_dir.to_str().unwrap());
+        println!("cargo:rustc-link-lib=static={}", LIB_NAME);
+        println!("cargo:rustc-cfg=feature=\"{}\"", HAS_LIB_FEATURE);
+    } else {
+        if build_lib(out_dir, OCAML_DIR) {
+            // Link with generated library.
+            println!("cargo:rustc-link-search={}", out_dir);
+            println!("cargo:rustc-link-lib=static={}", LIB_NAME);
+            println!("cargo:rustc-cfg=feature=\"{}\"", HAS_LIB_FEATURE);
+        } else {
+            // Build the Rust library and fail at runtime. The library uses the
+            // `no-tools` feature flag to implement the library with a panic.
+            println!(
+                "cargo:warning=wasm-spec-interpreter cannot build the \
+            OCaml library and will fail at runtime; see README.md for required \
+            dependencies."
+            );
+        }
+    }
+}
+
+fn build_lib(out_dir: &str, ocaml_dir: &str) -> bool {
+    if !has_build_tools() {
+        println!(
+            "cargo:warning=wasm-spec-interpreter is missing build tools; \
+            see README.md for required tools."
+        );
```

Comment:
Instead of boolean checks plus a `println!` I think it's safe to remove this and just use `cfg!(fuzzing)` to guard this `make`. If the `fuzzing` cfg is set then we should actually build the library and assert that the build succeeds.

Reply:
Basically unless `cfg(fuzzing)` is set I think this build script should be a noop, intentionally. With `cfg(fuzzing)` set, though, it would do everything it does now and assert that it succeeds.

Reply:
When I implemented this I ran into {errors while building the fuzz targets](https://github.com/bytecodealliance/wasmtime/runs/3243864078?check_suite_focus=true#step:7:127). I think what is happening is that `cfg!(fuzzing)` is not set when `build.rs` is run but _is_ set when the actual crate is compiled (hence the `compile_error!`). How is that possible?

Reply:
Ok, so I just removed the `if cfg!(fuzzing) ...` guard in the `build.rs` since it doesn't look like that will work.

Reply:
Ah ok I think in that case we'll want to go the route putting the ocaml bits and pieces behind a feature and activate the feature with a `target`-specific dependency in the fuzzing crate which enables the feature.

I think it's important to not print warnings when the build fails, because no one ever looks at the build logs for oss-fuzz so if warnings get printed there we'll never know.

Reply:
Ok, after examining what Cargo prints on failure I realize that the warnings are unnecessary: now that `build.rs` correctly panics when `make` fails Cargo will print out enough information to troubleshoot with. So I removed the warnings and have refactored the crate to use two features:
 - `build-libinterpret` is what we turn on during fuzzing and we assume the toolchain is enabled
 - `has-libinterpret` is turned on by `build.rs` to indicate that the OCaml library was successfully built and we should compile the non-panicking `interpret` function